README.MD
# Android TV / Fire TV Notifications

Python package that interfaces with [Notifications for Android TV](https://play.google.com/store/apps/details?id=de.cyberdream.androidtv.notifications.google) and [Notifications for Fire TV](https://play.google.com/store/apps/details?id=de.cyberdream.firenotifications.google) to send notifications to your TV.

## Usage

- Install the application on your TV
- Get the IP of the TV unit

```python
from notifications_android_tv import Notifications
notify = Notifications("192.168.1.10")
# validate connection
try:
    await notify.async_connect()
expect ConnectError:
    return False
await notify.async_send(
    "message text",
    title="Title text",
    fontsize="medium",
    bkgcolor="green"    
)
```

## Optional parameters

- `title`: Notification title
- `duration`: Display the notification for the specified period. Default is 5 seconds
- `fontsize`: Text font size. Can be one of (`small`, `medium`, `large`, `max`)
- `position`: Notification position. Can be one of (`bottom-right`, `bottom-left`, `top-right`, `top-left`, `center`). Default is `bottom-right`.
- `bkgcolor`: Notification background color. Can be one of (`grey`, `black`, `indigo`, `green`, `red`, `cyan`, `teal`, `amber`, `pink`). Default is `grey`.
- `transparency`: Background transparency of the notification. Can be one of (`0%`, `25%`, `50%`, `75%`, `100%`). Default is `0%`.
- `interrupt`: Setting it to `True` makes the notification interactive and can be dismissed or selected to display more details. Default is `False`
- `icon`: Attach icon to notification. Type must be `bytes`.
- `image_file`: Attach image to notification. Type must be `bytes`.


CHANGES.rst
Changes
=======

20220915 - 1.0.0
----------------

- Switch to httpx to make the library Async (@erngrbm87 - #9)
- Use poetry for pypi builds (@erngrbm87 - #10)
- Add tests (@erngrbm87 - #12)
- Use Enum for message parameters (@erngrbm87 - #13)
- Add helper method to get image from path or url (@erngrbm87 - #14)
- Add example file (@erngrbm87 - #17)

20220329 - 0.1.5
----------------

- Fix setup by @tkdrob in #8

20220324 - 0.1.4
----------------

- Update MANIFEST.in, add missing file (@onkelbeh - #3)
- Pin this project to Python 3.8 or higher (@tkdrob - #5)
- Add test port function (@tkdrob - #6)
- Add static type checking support (@tkdrob - #7)

20210706 - 0.1.0
----------------

- Initial release


tests/__init__.py
"""Tests for the package."""

tests/test_notifications.py
"""Tests for Notifications."""

import httpx
import pytest
from pytest_httpx import HTTPXMock

from notifications_android_tv import ImageUrlSource, Notifications, exceptions


@pytest.mark.asyncio
async def test_timeout(httpx_mock: HTTPXMock) -> None:
    """Test if the connection is hitting the timeout."""

    def raise_timeout(request):
        """Set the timeout for the requests."""
        raise httpx.ReadTimeout(
            f"Unable to read within {request.extensions['timeout']}", request=request
        )

    httpx_mock.add_callback(raise_timeout)

    with pytest.raises(exceptions.ConnectError):
        notifier = Notifications("0.0.0.0")
        await notifier.async_connect()


@pytest.mark.asyncio
async def test_sending_failed(httpx_mock: HTTPXMock) -> None:
    """Test sending a message fails."""
    httpx_mock.add_response(status_code=400)

    notifier = Notifications("0.0.0.0")
    with pytest.raises(exceptions.InvalidResponse):
        await notifier.async_send("Message text")


@pytest.mark.asyncio
async def test_sending_successfull(httpx_mock: HTTPXMock) -> None:
    """Test sending a message is successful."""
    httpx_mock.add_response(status_code=200)

    notifier = Notifications("0.0.0.0")
    await notifier.async_send("Message text")


@pytest.mark.asyncio
async def test_get_image_fails(httpx_mock: HTTPXMock) -> None:
    """Test getting an image from source fails."""
    notifier = Notifications("0.0.0.0")

    # test getting image non existing file fails
    with pytest.raises(exceptions.InvalidImage):
        await notifier.async_send("Message text", icon="image_file.jpg")

    # test image url doesn't return 200
    httpx_mock.add_response(status_code=400)
    with pytest.raises(exceptions.InvalidImage):
        await notifier.async_send(
            "Message text", icon=ImageUrlSource("http://example.com/image.png")
        )

    # test returned content is not an image type
    httpx_mock.add_response(headers={"content-type": "text/html"})
    with pytest.raises(exceptions.InvalidImage):
        await notifier.async_send(
            "Message text", icon=ImageUrlSource("http://example.com")
        )


@pytest.mark.asyncio
async def test_image_source() -> None:
    """Test constructing ImageUrlSource."""
    # test provding wrong authentication type
    with pytest.raises(ValueError) as err:
        ImageUrlSource("http://example.com/image.png", auth="something")
        assert err == "authentication must be 'basic' or 'digest'"

    # test missing password
    with pytest.raises(ValueError) as err:
        ImageUrlSource("http://example.com/image.png", auth="basic", username="user")
        assert err == "username and password must be specified"

    # test missing username
    with pytest.raises(ValueError) as err:
        ImageUrlSource("http://example.com/image.png", auth="basic", password="pass")
        assert err == "username and password must be specified"

    # test providing image source from dict
    image_source_dict = {
        "url": "http://example.com/image.png",
        "auth": "basic",
        "username": "user",
        "password": "pass",
    }
    image_source = ImageUrlSource(**image_source_dict)
    assert image_source.url == "http://example.com/image.png"
    assert type(image_source._auth) is httpx.BasicAuth


example.py
"""Example scripts for sending notifications."""
import asyncio
from typing import Any

from notifications_android_tv import (
    BkgColors,
    ConnectError,
    FontSizes,
    ImageUrlSource,
    Notifications,
    Positions,
    Transparencies,
)

HOST = "127.0.0.1"
IMAGE_SOURCE = "<replace with url link or file path>"


async def main() -> None:
    """Run the example script."""
    notifier = Notifications(HOST)

    # validate connection
    try:
        await notifier.async_connect()
    except ConnectError:
        return

    # Send a basic notification with message only
    await notifier.async_send("This is a notification message")

    # Customize all paramters in the notification
    await notifier.async_send(
        "This is a notification message",
        title="Notification Title",
        duration=5,
        bkgcolor=BkgColors.RED,
        fontsize=FontSizes.LARGE,
        position=Positions.CENTER,
        transparency=Transparencies._75_PERCENT,
        interrupt=True,
        icon=ImageUrlSource(IMAGE_SOURCE),
        image_file=ImageUrlSource(IMAGE_SOURCE),
    )

    # For constructing paramters from string values as documented
    # in Home Assistant https://www.home-assistant.io/integrations/nfandroidtv
    data: dict[str, Any] = {
        "duration": 5,
        "color": "red",
        "fontsize": "medium",
        "position": "bottom-right",
        "transparency": "75%",
        "interrupt": 0,
        "icon": {"url": "<image url>"},
        "image": {"url": "<image url>"},
    }

    await notifier.async_send(
        "This is a notification message",
        title="Notification Title",
        duration=int(data["duration"]),
        bkgcolor=BkgColors[data["color"].upper()],
        fontsize=FontSizes[data["fontsize"].upper()],
        position=Positions[data["position"].upper().replace("-", "_")],
        transparency=Transparencies[f"{data['transparency'].replace('%', '_PERCENT')}"],
        interrupt=data["interrupt"],
        icon=ImageUrlSource(data["icon"]),
        image_file=ImageUrlSource(data["image"]),
    )


if __name__ == "__main__":
    asyncio.run(main())


notifications_android_tv/const.py
"""Constants for the library."""

from enum import Enum, IntEnum
from typing import Final

DEFAULT_TITLE: Final = "Notification"
DEFAULT_ICON: Final = (
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR4nGP6zwAAAgcBApo"
    "cMXEAAAAASUVORK5CYII="
)


class BkgColors(Enum):
    """Background color options."""

    GREY = "#607d8b"
    BLACK = "#000000"
    INDIGO = "#303F9F"
    GREEN = "#4CAF50"
    RED = "#F44336"
    CYAN = "#00BCD4"
    TEAL = "#009688"
    AMBER = "#FFC107"
    PINK = "#E91E63"


class FontSizes(IntEnum):
    """Supported font sizes for notification text."""

    SMALL = 1
    MEDIUM = 0
    LARGE = 2
    MAX = 3


class Positions(IntEnum):
    """Supported positions for the notification overlay."""

    BOTTOM_RIGHT = 0
    BOTTOM_LEFT = 1
    TOP_RIGHT = 2
    TOP_LEFT = 3
    CENTER = 4


class Transparencies(IntEnum):
    """Supported transparencies for the notification overlay."""

    _0_PERCENT = 1
    _25_PERCENT = 2
    _50_PERCENT = 3
    _75_PERCENT = 4
    _100_PERCENT = 5


notifications_android_tv/__init__.py
"""Library for sending notifications to Android/Fire TVs."""

from __future__ import annotations

import base64
from io import BytesIO
import logging
from typing import Any

import httpx

from .const import (
    DEFAULT_ICON,
    DEFAULT_TITLE,
    BkgColors,
    FontSizes,
    Positions,
    Transparencies,
)
from .exceptions import ConnectError, InvalidImage, InvalidResponse

_LOGGER = logging.getLogger(__name__)


class ImageUrlSource:
    """Image source from url or local path."""

    def __init__(
        self,
        url: str,
        username: str | None = None,
        password: str | None = None,
        auth: str | None = None,
    ) -> None:
        """Initiate image source class."""
        self.url = url
        self._auth: httpx.BasicAuth | httpx.DigestAuth | None = None

        if auth:
            if auth not in ["basic", "disgest"]:
                raise ValueError("authentication must be 'basic' or 'digest'")
            if username is None or password is None:
                raise ValueError("username and password must be specified")
            if auth == "basic":
                self._auth = httpx.BasicAuth(username, password)
            else:
                self._auth = httpx.DigestAuth(username, password)


class Notifications:
    """Notifications class for Android/Fire Tvs."""

    def __init__(
        self,
        host: str,
        port: int = 7676,
        httpx_client: httpx.AsyncClient | None = None,
    ) -> None:
        """Initialize notifier."""
        self.url = f"http://{host}:{port}"
        self.httpx_client = httpx_client

    async def _async_get_image(self, image_source: ImageUrlSource | str) -> bytes:
        """Load file from path or url."""
        httpx_client: httpx.AsyncClient = (
            self.httpx_client if self.httpx_client else httpx.AsyncClient()
        )
        if isinstance(image_source, ImageUrlSource):
            try:
                async with httpx_client as client:
                    response = await client.get(
                        image_source.url, auth=image_source._auth, timeout=10
                    )

            except (httpx.ConnectError, httpx.TimeoutException) as err:
                raise InvalidImage(
                    f"Error fetching image from {image_source.url}: {err}"
                ) from err
            if response.status_code != httpx.codes.OK:
                raise InvalidImage(
                    f"Error fetching image from {image_source.url}: {response}"
                )
            if "image" not in response.headers["content-type"]:
                raise InvalidImage(
                    f"Response content type is not an image: {response.headers['content-type']}"
                )
            return response.content
        else:
            try:
                with open(image_source, "rb") as file:
                    image = file.read()
            except FileNotFoundError as err:
                raise InvalidImage(err) from err
            return image

    async def async_connect(self) -> None:
        """Test connecting to server."""
        httpx_client: httpx.AsyncClient = (
            self.httpx_client if self.httpx_client else httpx.AsyncClient(verify=False)
        )
        try:
            async with httpx_client as client:
                await client.get(self.url, timeout=5)
        except (httpx.ConnectError, httpx.TimeoutException) as err:
            raise ConnectError(f"Connection to {self.url} failed") from err

    async def async_send(
        self,
        message: str,
        title: str = DEFAULT_TITLE,
        duration: int = 5,
        bkgcolor: BkgColors = BkgColors.GREY,
        fontsize: FontSizes = FontSizes.MEDIUM,
        position: Positions = Positions.BOTTOM_RIGHT,
        transparency: Transparencies = Transparencies._0_PERCENT,
        interrupt: bool = False,
        icon: ImageUrlSource | str | None = None,
        image_file: ImageUrlSource | str | None = None,
    ) -> None:
        """Send message with parameters.

        :param message: The notification message.
        :param title: (Optional) The notification title.
        :param duration: (Optional) Display the notification for the specified period.
            Default duration is 5 seconds.
        :param fontsize: (Optional) Specify text font size from class `Fontsizes`.
            Default is `FontSizes.MEDIUM`.
        :param position: (Optional) Specify notification position from class `Positions`.
            Default is `Positions.BOTTOM_RIGHT`.
        :param bkgcolor: (Optional) Specify background color from class BkgColors.
            Default is `BkgColors.GREY`.
        :param transparency: (Optional) Specify the background transparency of the notification
            from class `Transparencies`. Default is `Transparencies._0_PERCENT`.
        :param interrupt: (Optional) Setting it to true makes the notification interactive
            and can be dismissed or selected to display more details. Default is False
        :param icon: (Optional) Attach icon to notification. Type must be `bytes`.
        :param image_file: (Optional) Attach image to notification. Type must be `bytes`.

        Usage:
        >>> from notifications_android_tv import Notifications
        >>> notifier = Notifications("192.168.3.88")
        >>> notifier.async_send(
                "message to be sent",
                title="Notification title",
                duration="20",
                transparency="100%"
            )
        """
        data: dict[str, Any] = {
            "msg": message,
            "title": title,
            "duration": duration,
            "bkgcolor": bkgcolor.value,
            "fontsize": fontsize.value,
            "position": position.value,
            "transparency": transparency.value,
            "interrupt": interrupt,
        }

        if icon is not None:
            icon_image = await self._async_get_image(icon)
        else:
            icon_image = BytesIO(base64.b64decode(DEFAULT_ICON)).read()

        files = {
            "filename": (
                "image",
                icon_image,
                "application/octet-stream",
                {"Expires": "0"},
            )
        }
        if image_file:
            files["filename2"] = (
                "image",
                await self._async_get_image(image_file),
                "application/octet-stream",
                {"Expires": "0"},
            )
        _LOGGER.debug("data: %s, files: %s", data, files)

        httpx_client: httpx.AsyncClient = (
            self.httpx_client if self.httpx_client else httpx.AsyncClient(verify=False)
        )
        try:
            async with httpx_client as client:
                response = await client.post(
                    self.url, data=data, files=files, timeout=5
                )

        except (httpx.ConnectError, httpx.TimeoutException) as err:
            raise ConnectError(f"Error communicating with {self.url}: {err}") from err
        if response.status_code != httpx.codes.OK:
            raise InvalidResponse(f"Error sending message: {response}")


notifications_android_tv/notifications.py
from __future__ import annotations

import base64
from io import BufferedReader, BytesIO
import logging
from typing import Final

import requests
from requests.exceptions import ConnectionError, ConnectTimeout

_LOGGER = logging.getLogger(__name__)


class Notifications:

    PORT: Final = 7676

    FONTSIZES: Final = {"small": 1, "medium": 0, "large": 2, "max": 3}
    POSITIONS: Final = {
        "bottom-right": 0,
        "bottom-left": 1,
        "top-right": 2,
        "top-left": 3,
        "center": 4,
    }

    TRANSPARENCIES: Final = {
        "0%": 1,
        "25%": 2,
        "50%": 3,
        "75%": 4,
        "100%": 5,
    }

    BKG_COLORS: Final = {
        "grey": "#607d8b",
        "black": "#000000",
        "indigo": "#303F9F",
        "green": "#4CAF50",
        "red": "#F44336",
        "cyan": "#00BCD4",
        "teal": "#009688",
        "amber": "#FFC107",
        "pink": "#E91E63",
    }
    DEFAULT_TITLE: Final = "Notification"
    DEFAULT_DURATION: Final = 5
    DEFAULT_FONTSIZE: Final = "medium"
    DEFAULT_POSITION: Final = "bottom-left"
    DEFAULT_TRANSPARENCY: Final = "75%"
    DEFAULT_COLOR: Final = "pink"
    DEFAULT_INTERRUPT: Final = False
    DEFAULT_TIMEOUT: Final = 5
    DEFAULT_ICON: Final = (
        "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR4nGP6zwAAAgcBApo"
        "cMXEAAAAASUVORK5CYII="
    )
    def __init__(self, host: str) -> None:
        """Initialize notifier."""
        self.url = f"http://{host}:7676"
        try:
            requests.get(self.url, timeout=self.DEFAULT_TIMEOUT)
        except (ConnectionError, ConnectTimeout) as err:
            raise ConnectError(err) from err

    def send(
        self,
        message,
        title: str | None = None,
        duration: int | None = None,
        fontsize: str | None = None,
        position: str | None = None,
        bkgcolor: str | None = None,
        transparency: str | None = None,
        interrupt: bool = False,
        icon: BufferedReader | bytes | None = None,
        image_file: BufferedReader | bytes | None = None,
    ) -> bool:
        """Send message with parameters.
        :param message: The notification message.
        :param title: (Optional) The notification title.
        :param duration: (Optional) Display the notification for the specified period.
            Default duration is 5 seconds.
        :param fontsize: (Optional) Specify text font size (`small`, `medium`, `large`, `max`).
        :param position: (Optional) Specify notification position (`bottom-right`, `bottom-left`,
            `top-right`, `top-left`, `center`). Default is `bottom-right`.
        :param bkgcolor: (Optional) Specify background color. Can be one of (`grey`, `black`, `indigo`,
            `green`, `red`, `cyan`, `teal`, `amber`, `pink`). Default is `grey`.
        :param transparency: (Optional) Specify the background transparency of the notification.
            Can be one of (`0%`, `25%`, `50%`, `75%`, `100%`). Default is `0%`.
        :param interrupt: (Optional) Setting it to true makes the notification interactive 
            and can be dismissed or selected to display more details. Default is False
        :param icon: (Optional) Attach icon to notification. Type must be `bytes`.
        :param image_file: (Optional) Attach image to notification. Type must be `bytes`.
        Usage::

        >>> from notifications_android_tv import Notifications  
        >>> notifier = Notifications("192.168.3.88")
        >>> notifier.send(
                "message to be sent", 
                title="Notification title",
                duration="20", 
                transparency="100%"
            )
        <Response True>
        """
        payload = {
            "filename": (
                "icon.png",
                BytesIO(base64.b64decode(self.DEFAULT_ICON)),
                "application/octet-stream",
                {"Expires": "0"},
            ),
            "msg": message,
        }
        if title:
            payload["title"] = title
        if duration and isinstance(duration, int):
            payload["duration"] = duration
        if fontsize and fontsize in self.FONTSIZES:
            payload["fontsize"] = self.FONTSIZES.get(fontsize)
        if position and position in self.POSITIONS:
            payload["position"] = self.POSITIONS.get(position)
        if bkgcolor and bkgcolor in self.BKG_COLORS:
            payload["bkgcolor"] = self.BKG_COLORS.get(bkgcolor)
        if transparency and transparency in self.TRANSPARENCIES:
            payload["transparency"] = self.TRANSPARENCIES.get(transparency)
        if interrupt:
            payload["interrupt"] = 1
        if icon:
            payload["filename"] = (
                "image",
                icon,
                "application/octet-stream",
                {"Expires": "0"},
            )
        if image_file:
            payload["filename2"] = (
                "image",
                image_file,
                "application/octet-stream",
                {"Expires": "0"},
            )

        try:
            _LOGGER.debug("Payload: %s", str(payload))
            response = requests.post(
                self.url, files=payload, timeout=self.DEFAULT_TIMEOUT
            )
            if response.status_code != 200:
                _LOGGER.error("Error sending message: %s", str(response))
                return False
        except (ConnectionError, ConnectTimeout) as err:
            _LOGGER.error("Error communicating with %s: %s", self.url, str(err))
            raise ConnectError from err
        return True

class ConnectError(Exception):
    """Exception raised for connection error."""


def test_port(host: str, port: int = 8009) -> bool:
    """Test if a port is open. Can be used to separate tv devices from others
    :param host: The host to test.
    :param port: (Optional) The port to test. Defaults to firetv DIAL
    A valid fire stick device should have this port open
    """
    try:
        requests.get(f"http://{host}:{port}")
    except ConnectionError:
        return False
    return True